radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+    TrackEverything,
+    /// Track everything from these peers, and nothing else.
+    TrackPeers(Vec<PeerId>),
+    /// Track the specified URNs.
+    TrackUrns(Vec<RadUrn>),
+}
+
+impl Mode {
+    fn is_trackable(&self, peer: &PeerId, urn: &RadUrn) -> bool {
+        match self {
+            Mode::TrackEverything => true,
+            Mode::TrackUrns(ref urns) => urns.contains(urn),
+            Mode::TrackPeers(ref peers) => peers.contains(peer),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct NodeConfig {
+    /// Address to listen to for new connections.
+    pub listen_addr: SocketAddr,
+    /// Operational mode. Determines the tracking rules for this seed node.
+    pub mode: Mode,
+    /// Radicle root path.
+    pub root: Option<PathBuf>,
+}
+
+impl Default for NodeConfig {
+    fn default() -> Self {
+        Self {
+            listen_addr: ([0, 0, 0, 0], 0).into(),
+            mode: Mode::TrackEverything,
+            root: None,
+        }
+    }
+}
+
+pub struct Node {
+    config: NodeConfig,
+}
+
+impl Node {
+    pub fn new(config: NodeConfig) -> Result<Self, Error> {
+        Ok(Node { config })
+    }
+
+    pub async fn run(self) -> Result<(), Error> {
+        use futures::stream::StreamExt;
+
+        let paths = if let Some(root) = self.config.root {
+            paths::Paths::from_root(root)?
+        } else {
+            paths::Paths::new()?
+        };
+        let gossip_params = Default::default();
+        let seeds: Vec<(PeerId, SocketAddr)> = vec![];
+        let disco = discovery::Static::new(seeds);
+        let signer = Signer::new();
+        let config = PeerConfig {
+            signer,
+            paths,
+            listen_addr: self.config.listen_addr,
+            gossip_params,
+            disco,
+        };
+
+        let peer = config.try_into_peer().await?;
+
+        let (api, future) = peer.accept()?;
+        let mut events = api.protocol().subscribe().await;
+        let mode = self.config.mode;
+
+        // Spawn the background peer thread.
+        tokio::spawn(future);
+
+        // Start by tracking specified projects if we need to.
+        if let Mode::TrackUrns(urns) = &mode {
+            for urn in urns {
+                let mut peers = api.providers(urn.clone()).await;
+                while let Some(peer) = peers.next().await {
+                    Node::track_project(&api, urn, &peer)?;
+                }
+            }
+        }
+
+        // Listen on gossip events. As soon as a peer announces something of interest,
+        // we check if we should track it.
+        while let Some(event) = events.next().await {
+            match event {
+                ProtocolEvent::Gossip(gossip::Info::Has(gossip::Has { provider, val })) => {
+                    let urn = &val.urn;
+                    let peer_id = &provider.peer_id;
+
+                    if mode.is_trackable(peer_id, urn) {
+                        Node::track_project(&api, urn, &provider)?;
```

Comment:
I suppose you don't want to crash if that fails, just spit out something in the logs.